opt: Factor constraint code into generalized library

andy-kimball opened the pull request:
The current constraints code is special-cased to locate applicable
indexes. This PR factors out some of the core functionality of that
code into a standalone package. In the future, we may use this
library to:
  - infer non-null columns from filter expressions
  - prune ranges
  - select indexes
  - push down filter conditions with equivalent columns into joins

The library itself defines key, span, constraint, and constraint
set. A key has one or more datum values. A span has a start and end
key, and can have inclusive or exclusive boundaries. A constraint
is a disjunction of spans that describe the domain of a list of
columns. And constraint set is a conjunction of constraints, with
each constraint constraining a different list of columns.

Constraint sets will be inferred from filter expression trees. They
can be combined together with Intersect or Union operators to infer
new constraint sets from component sets. A constraint is carefully
coded to only allocate a single object in the common case of one
constraint, one column, and one span (i.e. the constraint set for
@1 = 5 is a single allocation).

Together, these abstractions and operators will allow us to
efficiently compute and store a useful "summary" of the ranges of
data that the expression may access.

Release note: None

andy-kimball committed:
opt: Factor constraint code into generalized library

The current constraints code is special-cased to locate applicable
indexes. This PR factors out some of the core functionality of that
code into a standalone package. In the future, we may use this
library to:
  - infer non-null columns from filter expressions
  - prune ranges
  - select indexes
  - push down filter conditions with equivalent columns into joins

The library itself defines key, span, constraint, and constraint
set. A key has one or more datum values. A span has a start and end
key, and can have inclusive or exclusive boundaries. A constraint
is a disjunction of spans that describe the domain of a list of
columns. And constraint set is a conjunction of constraints, with
each constraint constraining a different list of columns.

Constraint sets will be inferred from filter expression trees. They
can be combined together with Intersect or Union operators to infer
new constraint sets from component sets. A constraint is carefully
coded to only allocate a single object in the common case of one
constraint, one column, and one span (i.e. the constraint set for
@1 = 5 is a single allocation).

Together, these abstractions and operators will allow us to
efficiently compute and store a useful "summary" of the ranges of
data that the expression may access.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/22768)
<!-- Reviewable:end -->

RaduBerinde commented:
Looks good so far, will have to make another pass.

I will just say that most of the changes from the conceptual model I used for index constraints (mainly using EMPTY and INF) come down to preference, I'm not convinced there are any significant practical advantages that justify the divergence (but what's done is done).

---

Review status: 0 of 10 files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[pkg/sql/opt/constraint/constraint.go, line 39 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5ReuN40DER-XqotdCu:-L5ReuN40DER-XqotdCv:b-dn4r89) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d376bf8f7b2c1a04cc67c9532390a547c78420a/pkg/sql/opt/constraint/constraint.go#L39)):*
> ```Go
> //
> // A few examples:
> //  - a constraint on @1 > 1: a single span             (/@1 (/1 - ∞])
> ```

Is there an easy way to type these unicode characters? Using them extensively in comments will get tedious..

---

*[pkg/sql/opt/constraint/constraint.go, line 41 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5RgDDR5Y4mrc-4vcfi:-L5RgDDR5Y4mrc-4vcfj:bc4y8eu) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d376bf8f7b2c1a04cc67c9532390a547c78420a/pkg/sql/opt/constraint/constraint.go#L41)):*
> ```Go
> //  - a constraint on @1 > 1: a single span             (/@1 (/1 - ∞])
> //  - a constraint on @1 = 1 AND @2 >= 1: a single span (/@1/@2 [/1/1 - /1])
> //  - a constraint on @1 < 5 OR @1 > 10: multiple spans (/@1 [Ø - /5) (10 - ∞))
> ```

There is an asymmetry between EMPTY and INF, wouldn't -INF instead of EMPTY make more sense?

---

*[pkg/sql/opt/constraint/constraint.go, line 113 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5RlHJb7MeBIxn90izf:-L5RlHJb7MeBIxn90izg:b-akw4yy) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d376bf8f7b2c1a04cc67c9532390a547c78420a/pkg/sql/opt/constraint/constraint.go#L113)):*
> ```Go
> // of spans).
> func (c *Constraint) Span(nth int) *Span {
> 	// There's always at least one span.
> ```

If there's always at least one span, why do we need `c.spanCount`? (instead of 1+len(c.otherSpans)`

---

*[pkg/sql/opt/constraint/constraint_set.go, line 66 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5RhA00EdyUtoKm9qOp:-L5RhA00EdyUtoKm9qOq:b24zjae) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d376bf8f7b2c1a04cc67c9532390a547c78420a/pkg/sql/opt/constraint/constraint_set.go#L66)):*
> ```Go
> 
> 	// len is the number of constraints in the set.
> 	len int32
> ```

[nit] `len` is not a great name, it is a built-in (if nothing else, syntax highlighting makes it look weird)

---

*[pkg/sql/opt/constraint/constraint_set.go, line 131 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5Rm6kDSGrD6yfaYnAT:-L5Rm6kDSGrD6yfaYnAU:b-19e3cl) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d376bf8f7b2c1a04cc67c9532390a547c78420a/pkg/sql/opt/constraint/constraint_set.go#L131)):*
> ```Go
> 
> 	index := 0
> 	len := c.Length()
> ```

[nit] `len` shadows the built-in

---

*[pkg/sql/opt/constraint/key.go, line 46 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5RfuLVCsDwdqXQQfXb:-L5RfuLVCsDwdqXQQfXc:b-ymupje) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d376bf8f7b2c1a04cc67c9532390a547c78420a/pkg/sql/opt/constraint/key.go#L46)):*
> ```Go
> //   exclusive start = invalid (prevented by Span.Set)
> //   inclusive end   = ∞
> //   exclusive end   = Ø
> ```

this is only used in empty spans right? (maybe mention that here)

---

*[pkg/sql/opt/constraint/key.go, line 53 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5TsSvfWRSecD-pAQl8:-L5TsSvfWRSecD-pAQl9:bpkhrw8) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d376bf8f7b2c1a04cc67c9532390a547c78420a/pkg/sql/opt/constraint/key.go#L53)):*
> ```Go
> 
> // Key is a composite of N typed datum values that are ordered from most
> // significant to least significant.
> ```

"The datum values correspond to a set of columns; it is the responsibility of the calling code to keep track of them."

---

*[pkg/sql/opt/constraint/key.go, line 206 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5TrvhrbEANJAVYH4DY:-L5TrvhrbEANJAVYH4DZ:b-rmugzm) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d376bf8f7b2c1a04cc67c9532390a547c78420a/pkg/sql/opt/constraint/key.go#L206)):*
> ```Go
> 
> 		default:
> 			vals := make(tree.Datums, len(k.otherVals)+1+len(l.otherVals))
> ```

I think we have too many cases here. Doesn't this piece of code work just fine if one of the lengths in 1? Btw I'd just reuse the result from l.Length(), k.Length() instead of recalculating (1+len()).

---

*[pkg/sql/opt/constraint/span.go, line 31 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5TtWkluXyf7ad6He8i:-L5TtWkluXyf7ad6He8j:bmjpd8z) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d376bf8f7b2c1a04cc67c9532390a547c78420a/pkg/sql/opt/constraint/span.go#L31)):*
> ```Go
> 
> const (
> 	// ExcludeBoundary indicates that the boundary does not include the start
> ```

This can be misleading (it sounds like it is about both endpoints). Myabe "ExcludeBoundary indicates that the boundary does not include the respective key"

---

*[pkg/sql/opt/constraint/span.go, line 63 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5TttzsHukbNy9TDd_i:-L5TttzsHukbNy9TDd_j:b3v4o5u) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d376bf8f7b2c1a04cc67c9532390a547c78420a/pkg/sql/opt/constraint/span.go#L63)):*
> ```Go
> 	end Key
> 
> 	// startExt is the key extension value (empty or infinity) that is used
> ```

I would have expected to use the `SpanBoundary` here, which is easier to think about than the "extension". We could represent the empty span as `(EMPTY - EMPTY)` and the zero value would still be the empty span (in fact, that choice seems cleaner than the somewhat arbitrary choice between `[empty - empty)` and `(empty - empty]`

---

*[pkg/sql/opt/constraint/span.go, line 75 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5TuvJbKKhPiRBaYQyl:-L5TuvJbKKhPiRBaYQym:brk8ss) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d376bf8f7b2c1a04cc67c9532390a547c78420a/pkg/sql/opt/constraint/span.go#L75)):*
> ```Go
> 
> // TrySet sets the boundaries of this span to the given values. If the start
> // boundary is >= the end boundary, TrySet returns false and does not make any
> ```

This can easily read as start key >= end key which is not accurate. Maybe just say "if the result would be an empty span"

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-:-L5TwId_gJ-FATX121rN:bcwy2et)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>  I like the updated version a lot!

---

Review status: 0 of 10 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/sql/opt/constraint/constraint_set.go, line 131 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5Rm6kDSGrD6yfaYnAT:-L5YwSIGapuMhD7_5-ZT:b-s087qx) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d376bf8f7b2c1a04cc67c9532390a547c78420a/pkg/sql/opt/constraint/constraint_set.go#L131)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

This one is still here.

---

*[pkg/sql/opt/constraint/key.go, line 206 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5TrvhrbEANJAVYH4DY:-L5YwlVq9d17ObQVZVQc:be5es6q) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d376bf8f7b2c1a04cc67c9532390a547c78420a/pkg/sql/opt/constraint/key.go#L206)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I changed up the lengths a bit. Does Go have any way of easily referencing the value being switched on? I just stored the lens in separate vars. I also don't see what you mean by having too many cases. What case(s) could I remove?
</blockquote></details>

You can define the var in the switch statement:`switch l := <something> {}` and then use `l`. This even works in type switches, and `l` will have the respective type in each branch.

I don't think we need the special cases for lengths of `1`, the general code should work just fine. I believe that the function can just be:
```go
if kLen == 0 { return l }
if lLen == 0 { return k }

vals := make(tree.Datums, kLen+lLen-1)
copy(vals, k.otherVals)
vals[kLen-1] = l.firstVal
copy(vals[kLen:], l.otherVals)
return Key{firstVal: k.firstVal, otherVals: vals}
```

---

*[pkg/sql/opt/constraint/key.go, line 27 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5Ywd8PbZreDO9anqvD:-L5Ywd8PbZreDO9anqvE:b4c9lyr) ([raw file](https://github.com/cockroachdb/cockroach/blob/a322fbd5ec6b635c686b79822ff22cddd24fb522/pkg/sql/opt/constraint/key.go#L27)):*
> ```Go
> // specify whether each key is conceptually suffixed with a value that sorts
> // before all values (ExtendLow) or with a value that sorts after all values
> // (ExtendHigh). This allows span boundaries to be compared with one another.
> ```

:thumbsup: 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-:-L5YyLqqonZ92KDwX3Gb:bxx5j0w)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
Review status: 0 of 10 files reviewed at latest revision, 5 unresolved discussions.

---

*[pkg/sql/opt/constraint/constraint_set.go, line 131 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5Rm6kDSGrD6yfaYnAT:-L5ZMQ8qNJ5aYJMjgx2d:b-d3mclt) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d376bf8f7b2c1a04cc67c9532390a547c78420a/pkg/sql/opt/constraint/constraint_set.go#L131)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

This one is still here.
</blockquote></details>

Whoops...fixed.

---

*[pkg/sql/opt/constraint/key.go, line 53 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5TsSvfWRSecD-pAQl8:-L5ZQ3SinWjAfdlQ15oi:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d376bf8f7b2c1a04cc67c9532390a547c78420a/pkg/sql/opt/constraint/key.go#L53)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

"The datum values correspond to a set of columns; it is the responsibility of the calling code to keep track of them."
</blockquote></details>

Done.

---

*[pkg/sql/opt/constraint/key.go, line 206 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5TrvhrbEANJAVYH4DY:-L5ZMzbaBn1fklHuiSbJ:b-2mhfem) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d376bf8f7b2c1a04cc67c9532390a547c78420a/pkg/sql/opt/constraint/key.go#L206)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

You can define the var in the switch statement:`switch l := <something> {}` and then use `l`. This even works in type switches, and `l` will have the respective type in each branch.

I don't think we need the special cases for lengths of `1`, the general code should work just fine. I believe that the function can just be:
```go
if kLen == 0 { return l }
if lLen == 0 { return k }

vals := make(tree.Datums, kLen+lLen-1)
copy(vals, k.otherVals)
vals[kLen-1] = l.firstVal
copy(vals[kLen:], l.otherVals)
return Key{firstVal: k.firstVal, otherVals: vals}
```
</blockquote></details>

I had tried using an assignment with the switch statement and got an error. Now I realize the parser needs a semicolon after the assignment to make it work: `switch l := <something>; {}`. Anyway, your version is better, not sure why I didn't see it in the first place. Changed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 10 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-:-L5Znto3ITu-qV9GGe88:bnfp4nl)*
<!-- Sent from Reviewable.io -->

rytaft commented:
Looks good so far, but this logic is very tricky. I've identified a few places where I think you can expand the comments a bit.  I've reviewed a couple of files - will come back and do the others in a bit.

---

Reviewed 2 of 5 files at r3.
Review status: 2 of 10 files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt/constraint/constraint.go, line 41 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5ikvuoQzT9RsHx-33j:-L5ikvupBJG0seUQYLFt:b-1zsygf) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4c5a20a88adfae4b2c389ec27c6da8cb643396/pkg/sql/opt/constraint/constraint.go#L41)):*
> ```Go
> //  - a constraint on @1 > 1: a single span             /@1: (/1 - ]
> //  - a constraint on @1 = 1 AND @2 >= 1: a single span /@1/@2: [/1/1 - /1]
> //  - a constraint on @1 < 5 OR @1 > 10: multiple spans /@1: [ - /5) (10 - )
> ```

Why is "[" used for -inf and ")" used for +inf? Why not be consistent? 

(edit: I also noticed in `constraint_set.go` you are using "]" for +inf)

---

*[pkg/sql/opt/constraint/constraint.go, line 115 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5inNeZa6dPwVrmdfhJ:-L5inNeZa6dPwVrmdfhK:bqgg6tx) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4c5a20a88adfae4b2c389ec27c6da8cb643396/pkg/sql/opt/constraint/constraint.go#L115)):*
> ```Go
> // constraints. If the merge results in an unconstrained span, then
> // tryUnionWith returns false. An unconstrained span should never be part of a
> // constraint, as then the constrained would not constrain anything, and should
> ```

constrained -> span  ?

---

*[pkg/sql/opt/constraint/constraint.go, line 168 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5irFUW5u879_JQ2V6a:-L5irFUW5u879_JQ2V6b:bj7fxij) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4c5a20a88adfae4b2c389ec27c6da8cb643396/pkg/sql/opt/constraint/constraint.go#L168)):*
> ```Go
> 			}
> 
> 			// If neither union succeeded (because spans are disjoint), then
> ```

This is a bit confusing since Constraint.tryUnionWith is doing something very different from Span.TryUnionWith. It might help to expand this comment a bit to point that out.

---

*[pkg/sql/opt/constraint/constraint_set.go, line 48 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5iv_Yi2AG499KbRoM7:-L5iv_Yi2AG499KbRoM8:bsnuq4r) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4c5a20a88adfae4b2c389ec27c6da8cb643396/pkg/sql/opt/constraint/constraint_set.go#L48)):*
> ```Go
> // A few examples:
> //  - @1 > 10
> //      (/@1 (/1 - ])
> ```

I think you meant `(/@1 (/10 - ])`

---

*[pkg/sql/opt/constraint/constraint_set.go, line 51 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5iw3_cElMapAYm0nIj:-L5iw3_cElMapAYm0nIk:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4c5a20a88adfae4b2c389ec27c6da8cb643396/pkg/sql/opt/constraint/constraint_set.go#L51)):*
> ```Go
> //
> //  - @1 > 10 AND @2 = 5
> //      (/@1 (/1 - ])
> ```

ditto

---

*[pkg/sql/opt/constraint/constraint_set.go, line 94 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5ix6e1ZOcIFc3YdxE7:-L5ix6e1ZOcIFc3YdxE8:bvg5crm) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4c5a20a88adfae4b2c389ec27c6da8cb643396/pkg/sql/opt/constraint/constraint_set.go#L94)):*
> ```Go
> 
> // Constraint returns the nth constraint in the set. Together with the
> // ConstraintCount method, Constraint allows iteration over the list of
> ```

There is no `ConstraintCount` method - did you mean `Length`?

---

*[pkg/sql/opt/constraint/constraint_set.go, line 95 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5ixqHeYE3gyx-H1r6C:-L5ixqHfaHdOtniKQ0Vp:b-hvl221) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4c5a20a88adfae4b2c389ec27c6da8cb643396/pkg/sql/opt/constraint/constraint_set.go#L95)):*
> ```Go
> // Constraint returns the nth constraint in the set. Together with the
> // ConstraintCount method, Constraint allows iteration over the list of
> // comstraints (since there is no method to return a slice of constraints).
> ```

comstraints -> constraints

---

*[pkg/sql/opt/constraint/constraint_set.go, line 211 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5j2eNWU9UsPrpVmmUR:-L5j0UJER17tuDKObo3-:b52667w) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4c5a20a88adfae4b2c389ec27c6da8cb643396/pkg/sql/opt/constraint/constraint_set.go#L211)):*
> ```Go
> 	for index < length && otherIndex < otherLength {
> 		// Skip past constraints that do not have corresponding constraint in
> 		// the other set.
> ```

The fact that you're skipping past constraints that don't match is a bit confusing since "union" makes me think the union should contain all constraints.  I understand why you're doing this after thinking about it for a bit, but it might help future maintainers if you expand this comment a bit.

---

*[pkg/sql/opt/constraint/constraint_set.go, line 260 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5j1uGqGHRjUOKOJuHZ:-L5j1uGqGHRjUOKOJuH_:buswd7w) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4c5a20a88adfae4b2c389ec27c6da8cb643396/pkg/sql/opt/constraint/constraint_set.go#L260)):*
> ```Go
> 		panic("correct capacity should have been set when otherConstraints was allocated")
> 	}
> 	s.otherConstraints = s.otherConstraints[:s.length-1]
> ```

This is kind of confusing since it looks like you're reducing the size of the slice, but in fact you're extending it. Maybe a comment would help reminding us that s.length is different from len(s.otherConstraints)?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-:-L5j6J7DpgO92mghMByD:b4vufht)*
<!-- Sent from Reviewable.io -->

rytaft commented:
Reviewed 2 of 5 files at r3.
Review status: 4 of 10 files reviewed at latest revision, 27 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt/constraint/constraint_set_test.go, line 16 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5jIfC72ShAisj26UKk:-L5jIfC72ShAisj26UKl:b10ca2m) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4c5a20a88adfae4b2c389ec27c6da8cb643396/pkg/sql/opt/constraint/constraint_set_test.go#L16)):*
> ```Go
> 
> package constraint
> ```

Nice tests!

---

*[pkg/sql/opt/constraint/constraint_set_test.go, line 53 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5jHfmKPMfldX94XCgZ:-L5jHfmKPMfldX94XCg_:b-dsq8vo) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4c5a20a88adfae4b2c389ec27c6da8cb643396/pkg/sql/opt/constraint/constraint_set_test.go#L53)):*
> ```Go
> 
> 	// Include constraint on multiple columns.
> 	// (@1, @2) > (10, 15)
> ```

I think this should be (@1, @2) >= (10, 15)

(Fix all the comments below too)

---

*[pkg/sql/opt/constraint/constraint_set_test.go, line 116 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5jJUzilYrwuZD42O5i:-L5jJUzilYrwuZD42O5j:bhvloc1) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4c5a20a88adfae4b2c389ec27c6da8cb643396/pkg/sql/opt/constraint/constraint_set_test.go#L116)):*
> ```Go
> 	test(t, &evalCtx, res, "/1: (/20 - /40]\n")
> 
> 	// Intersect with Contradiction (always contradiction).
> ```

You're not actually testing intersect with Contradiction - you're testing that two disjoint sets create a contradiction (also a good test!)

---

*[pkg/sql/opt/constraint/constraint_set_test.go, line 165 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5jFBlV_Mq0w7yIAXWX:-L5jFBlV_Mq0w7yIAXWY:b-rz9klg) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4c5a20a88adfae4b2c389ec27c6da8cb643396/pkg/sql/opt/constraint/constraint_set_test.go#L165)):*
> ```Go
> 
> 	// Union incompatible constraints (both are discarded).
> 	// (@1, @2) > (10, 15) AND @2 < 80
> ```

Shouldn't this be: `(@1, @2) > (10, 15) OR @2 < 80`?

---

*[pkg/sql/opt/constraint/constraint_set_test.go, line 204 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5jJqCb8XtMIpaAF-UG:-L5jJqCb8XtMIpaAF-UH:bnau1ds) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4c5a20a88adfae4b2c389ec27c6da8cb643396/pkg/sql/opt/constraint/constraint_set_test.go#L204)):*
> ```Go
> 		"/2: [ - /80]\n")
> 
> 	// Union with Unconstrained (always unconstrained).
> ```

What about also testing, for example, that `spGt20.Union(&evalCtx, spLe40)` is unconstrained?

---

*[pkg/sql/opt/constraint/constraint_set_test.go, line 222 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5jC-87ltLryVPJU5g5:-L5jC-87ltLryVPJU5g6:b-2760pn) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4c5a20a88adfae4b2c389ec27c6da8cb643396/pkg/sql/opt/constraint/constraint_set_test.go#L222)):*
> ```Go
> 	spLe40   Span // [ - /40]
> 	spLt80   Span // [ - /80)
> 	spEq80   Span // [/80 - /80)
> ```

should this be [/80 - /80]?

---

*[pkg/sql/opt/constraint/constraint_set_test.go, line 247 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5jBokRqXv81ttGeQOm:-L5jBokRqXv81ttGeQOn:b-2760pn) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4c5a20a88adfae4b2c389ec27c6da8cb643396/pkg/sql/opt/constraint/constraint_set_test.go#L247)):*
> ```Go
> 	data.spLt80.Set(evalCtx, EmptyKey, IncludeBoundary, key80, ExcludeBoundary)
> 
> 	// [/0 - /80)
> ```

should this be [/80 - /80]?

---

*[pkg/sql/opt/constraint/constraint_test.go, line 38 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5jKeY3f-jEOFz4s-TM:-L5jKeY3f-jEOFz4s-TN:b-l5lerm) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4c5a20a88adfae4b2c389ec27c6da8cb643396/pkg/sql/opt/constraint/constraint_test.go#L38)):*
> ```Go
> 		if actual != expected {
> 			format := "left: %s, right: %s, expected: %v, actual: %v"
> 			t.Errorf(format, left.String(), right.String(), expected, clone.String())
> ```

I think you should change `clone.String()` to `actual` 

---

*[pkg/sql/opt/constraint/constraint_test.go, line 57 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5jNFOw0UKovQcI4bdY:-L5jNFOw0UKovQcI4bdZ:b-jqu146) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4c5a20a88adfae4b2c389ec27c6da8cb643396/pkg/sql/opt/constraint/constraint_test.go#L57)):*
> ```Go
> 	test(t, &evalCtx, &data.c40to50, &data.c1to10, "/1: [/1 - /10] [/40 - /50]")
> 
> 	// Tight disjoint spans in each constraint.
> ```

How is this "tight"?

---

*[pkg/sql/opt/constraint/constraint_test.go, line 64 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5jQApN2fNmX2En7ATo:-L5jQApN2fNmX2En7ATp:bvkfvts) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4c5a20a88adfae4b2c389ec27c6da8cb643396/pkg/sql/opt/constraint/constraint_test.go#L64)):*
> ```Go
> 	var left, right Constraint
> 	left = data.c1to10
> 	_ = left.tryUnionWith(&evalCtx, &data.c20to30)
> ```

Why not have another helper function to call `tryUnionWith` and also check that the boolean output is correct?

---

*[pkg/sql/opt/constraint/constraint_test.go, line 118 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5jP2DsJix-tU9h2zZQ:-L5jP2DsJix-tU9h2zZR:b-l5lerm) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4c5a20a88adfae4b2c389ec27c6da8cb643396/pkg/sql/opt/constraint/constraint_test.go#L118)):*
> ```Go
> 		if actual != expected {
> 			format := "left: %s, right: %s, expected: %v, actual: %v"
> 			t.Errorf(format, left.String(), right.String(), expected, clone.String())
> ```

I think you should change `clone.String()` to `actual` 

---

*[pkg/sql/opt/constraint/constraint_test.go, line 175 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5jLA53DlOCTBWunAdX:-L5jLA53DlOCTBWunAdY:b-3cxi5l) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4c5a20a88adfae4b2c389ec27c6da8cb643396/pkg/sql/opt/constraint/constraint_test.go#L175)):*
> ```Go
> type constraintTestData struct {
> 	cLt10           Constraint // [Ø - /10)
> 	cGt20           Constraint // (/20 - ∞]
> ```

change this to [ - /10) and (/20 - ]

---

*[pkg/sql/opt/constraint/constraint_test.go, line 183 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5jMj0QPNtwnORtqHEa:-L5jMj0QPNtwnORtqHEb:bg87jfo) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4c5a20a88adfae4b2c389ec27c6da8cb643396/pkg/sql/opt/constraint/constraint_test.go#L183)):*
> ```Go
> 	c60to70         Constraint // (/60 - /70)
> 	cherryRaspberry Constraint // [/'cherry'/true - /'raspberry'/false)
> 	mangoStrawberry Constraint // [/'mango'/true - /'strawberry']
> ```

should be `/'mango'/false`

---

*[pkg/sql/opt/constraint/constraint_test.go, line 207 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5jLGwu4VA9XVZTvJcg:-L5jLGwu4VA9XVZTvJch:blvl4bl) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4c5a20a88adfae4b2c389ec27c6da8cb643396/pkg/sql/opt/constraint/constraint_test.go#L207)):*
> ```Go
> 	var span Span
> 
> 	// [Ø - /10)
> ```

change this to [ - /10)

---

*[pkg/sql/opt/constraint/constraint_test.go, line 211 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5jLHkdEukHeQjGXI3P:-L5jLHkdEukHeQjGXI3Q:b-hsmpib) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4c5a20a88adfae4b2c389ec27c6da8cb643396/pkg/sql/opt/constraint/constraint_test.go#L211)):*
> ```Go
> 	data.cLt10.init(1, &span)
> 
> 	// (/20 - ∞]
> ```

change this to (/20 - ]

---

*[pkg/sql/opt/constraint/constraint_test.go, line 243 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5jMH7Kd9qcoMT-s9_A:-L5jMH7Kd9qcoMT-s9_B:bg87jfo) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4c5a20a88adfae4b2c389ec27c6da8cb643396/pkg/sql/opt/constraint/constraint_test.go#L243)):*
> ```Go
> 	data.cherryRaspberry.initComposite([]opt.ColumnIndex{1, 2}, &span)
> 
> 	// [/'mango'/true - /'strawberry']
> ```

should be `/'mango'/false`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768)*
<!-- Sent from Reviewable.io -->

rytaft commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>

---

Reviewed 2 of 10 files at r1, 3 of 8 files at r2, 1 of 5 files at r3.
Review status: all files reviewed at latest revision, 34 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt/constraint/key.go, line 193 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5jTR-f9_cl6Pz0P9Rz:-L5jTR-f9_cl6Pz0P9S-:b-ytjvi6) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4c5a20a88adfae4b2c389ec27c6da8cb643396/pkg/sql/opt/constraint/key.go#L193)):*
> ```Go
> // maps to -1 because it sorts before all other values. ExtendHigh maps to 1
> // because it sorts after all other values.
> func (e KeyExtension) ToCmp() int {
> ```

Would be nice to move this up to the `KeyExtension` type definition so all the methods for the same type are together

---

*[pkg/sql/opt/constraint/span.go, line 45 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5jWpuM7Rf9IrowISDa:-L5jWpuM7Rf9IrowISDb:bq4i8rd) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4c5a20a88adfae4b2c389ec27c6da8cb643396/pkg/sql/opt/constraint/span.go#L45)):*
> ```Go
> // examples:
> //   @1 < 100                                          : ( - /100)
> //   @1 >= 100                                         : [/100 - )
> ```

Make the +inf and -inf square brackets

---

*[pkg/sql/opt/constraint/span.go, line 224 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5jbN2aUtppZ9i6aR66:-L5jbN2aUtppZ9i6aR67:bqdp90p) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4c5a20a88adfae4b2c389ec27c6da8cb643396/pkg/sql/opt/constraint/span.go#L224)):*
> ```Go
> 
> 	var cmp int
> 	if cmpStartKeys <= 0 {
> ```

Shouldn't this just be `<`, not `<=`?

---

*[pkg/sql/opt/constraint/span.go, line 229 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5jbItk6AfGS91WFQ3J:-L5jbItlW3Ps-AN9znPy:bv21dmd) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4c5a20a88adfae4b2c389ec27c6da8cb643396/pkg/sql/opt/constraint/span.go#L229)):*
> ```Go
> 		cmp = sp.end.Compare(evalCtx, other.start, sp.endExt(), other.startExt())
> 	} else {
> 		// This span is less, so see if there's any "space" after the end of
> ```

Should say this span is greater

---

*[pkg/sql/opt/constraint/span.go, line 230 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5jbYf9excwAj9jSpjr:-L5jbYf9excwAj9jSpjs:b-purip1) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4c5a20a88adfae4b2c389ec27c6da8cb643396/pkg/sql/opt/constraint/span.go#L230)):*
> ```Go
> 	} else {
> 		// This span is less, so see if there's any "space" after the end of
> 		// the other span and the start of this span.
> ```

...and before the start of...

---

*[pkg/sql/opt/constraint/span_test.go, line 331 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5jgIaOQAEaWyhHTB3W:-L5jgIaOQAEaWyhHTB3X:b6kyya6) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4c5a20a88adfae4b2c389ec27c6da8cb643396/pkg/sql/opt/constraint/span_test.go#L331)):*
> ```Go
> }
> 
> func TestSpanIntersect(t *testing.T) {
> ```

What about testing intersection with `Span{}`?

---

*[pkg/sql/opt/constraint/span_test.go, line 421 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5jfYGC5VXli5uyN9Ks:-L5jfYGC5VXli5uyN9Kt:b-fdiljk) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4c5a20a88adfae4b2c389ec27c6da8cb643396/pkg/sql/opt/constraint/span_test.go#L421)):*
> ```Go
> 	)
> 	testInt(t, &evalCtx, pear, pear2, "(/'pear'/5/NULL - /'pear'/10]")
> 	testInt(t, &evalCtx, pear, pear2, "(/'pear'/5/NULL - /'pear'/10]")
> ```

Did you mean to swap `pear` and `pear2` here?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-:-L5jh6a3IS7X5znYfu0a:b-jbsf8z)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
Review status: all files reviewed at latest revision, 34 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt/constraint/constraint.go, line 41 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5ikvuoQzT9RsHx-33j:-L5kzggXAV_zQb1Iq84a:b-fphlys) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4c5a20a88adfae4b2c389ec27c6da8cb643396/pkg/sql/opt/constraint/constraint.go#L41)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Why is "[" used for -inf and ")" used for +inf? Why not be consistent? 

(edit: I also noticed in `constraint_set.go` you are using "]" for +inf)
</blockquote></details>

I missed updating some of these comments, as it used to be `)` in some of these cases. They should now consistently be `]`, so if you notice any lingering `)`, point them out and I'll fix.

---

*[pkg/sql/opt/constraint/constraint.go, line 115 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5inNeZa6dPwVrmdfhJ:-L5l-Juy09hSB3HRNXpV:b2w3oju) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4c5a20a88adfae4b2c389ec27c6da8cb643396/pkg/sql/opt/constraint/constraint.go#L115)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

constrained -> span  ?
</blockquote></details>

Should be `constraint`. Fixed. We should never create a constraint that doesn't actually constrain anything.

---

*[pkg/sql/opt/constraint/constraint.go, line 168 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5irFUW5u879_JQ2V6a:-L5l-tB3fqpWaKB0Mf4f:b-4v7o0w) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4c5a20a88adfae4b2c389ec27c6da8cb643396/pkg/sql/opt/constraint/constraint.go#L168)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

This is a bit confusing since Constraint.tryUnionWith is doing something very different from Span.TryUnionWith. It might help to expand this comment a bit to point that out.
</blockquote></details>

You mean that the boolean return value indicates something very different? Both mean "the result of the union cannot be represented as a valid Span/Constraint". I added some commentary to that effect, but if you want to suggest something that further reduces the confusion, please do and I can add. I also changed the logic a bit to be more explicit about Unconstrained Span handling. Not allowing Unconstrained Constraints was something Radu felt was important, and that drives some of this logic.

---

*[pkg/sql/opt/constraint/constraint_set.go, line 48 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5iv_Yi2AG499KbRoM7:-L5lGzc21gFu6rpueGdR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4c5a20a88adfae4b2c389ec27c6da8cb643396/pkg/sql/opt/constraint/constraint_set.go#L48)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

I think you meant `(/@1 (/10 - ])`
</blockquote></details>

Done.

---

*[pkg/sql/opt/constraint/constraint_set.go, line 51 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5iw3_cElMapAYm0nIj:-L5lH-MDGtNngRfRsr8Y:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4c5a20a88adfae4b2c389ec27c6da8cb643396/pkg/sql/opt/constraint/constraint_set.go#L51)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done.

---

*[pkg/sql/opt/constraint/constraint_set.go, line 94 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5ix6e1ZOcIFc3YdxE7:-L5lH7TYMdYmryU8STWN:bktii9h) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4c5a20a88adfae4b2c389ec27c6da8cb643396/pkg/sql/opt/constraint/constraint_set.go#L94)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

There is no `ConstraintCount` method - did you mean `Length`?
</blockquote></details>

Yep, copy/paste issue. Done.

---

*[pkg/sql/opt/constraint/constraint_set.go, line 95 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5ixqHeYE3gyx-H1r6C:-L5lHA0PTUznhgILFZqO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4c5a20a88adfae4b2c389ec27c6da8cb643396/pkg/sql/opt/constraint/constraint_set.go#L95)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

comstraints -> constraints
</blockquote></details>

Done.

---

*[pkg/sql/opt/constraint/constraint_set.go, line 211 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5j2eNWU9UsPrpVmmUR:-L5lHJuJtneD7gBfz05z:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4c5a20a88adfae4b2c389ec27c6da8cb643396/pkg/sql/opt/constraint/constraint_set.go#L211)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

The fact that you're skipping past constraints that don't match is a bit confusing since "union" makes me think the union should contain all constraints.  I understand why you're doing this after thinking about it for a bit, but it might help future maintainers if you expand this comment a bit.
</blockquote></details>

Done.

---

*[pkg/sql/opt/constraint/constraint_set.go, line 260 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5j1uGqGHRjUOKOJuHZ:-L5lW7GcsqS1Zw5kHsPX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4c5a20a88adfae4b2c389ec27c6da8cb643396/pkg/sql/opt/constraint/constraint_set.go#L260)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

This is kind of confusing since it looks like you're reducing the size of the slice, but in fact you're extending it. Maybe a comment would help reminding us that s.length is different from len(s.otherConstraints)?
</blockquote></details>

Done.

---

*[pkg/sql/opt/constraint/constraint_set_test.go, line 53 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5jHfmKPMfldX94XCgZ:-L5lWUaVd-KFqseOUykX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4c5a20a88adfae4b2c389ec27c6da8cb643396/pkg/sql/opt/constraint/constraint_set_test.go#L53)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

I think this should be (@1, @2) >= (10, 15)

(Fix all the comments below too)
</blockquote></details>

Done.

---

*[pkg/sql/opt/constraint/constraint_set_test.go, line 116 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5jJUzilYrwuZD42O5i:-L5lXkNyFi-wCdt9Qeg7:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4c5a20a88adfae4b2c389ec27c6da8cb643396/pkg/sql/opt/constraint/constraint_set_test.go#L116)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

You're not actually testing intersect with Contradiction - you're testing that two disjoint sets create a contradiction (also a good test!)
</blockquote></details>

Fixed.

---

*[pkg/sql/opt/constraint/constraint_set_test.go, line 165 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5jFBlV_Mq0w7yIAXWX:-L5lY1QR6cFsYZ8Q0z_0:b-pkzycw) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4c5a20a88adfae4b2c389ec27c6da8cb643396/pkg/sql/opt/constraint/constraint_set_test.go#L165)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Shouldn't this be: `(@1, @2) > (10, 15) OR @2 < 80`?
</blockquote></details>

Yes, fixed.

---

*[pkg/sql/opt/constraint/constraint_set_test.go, line 204 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5jJqCb8XtMIpaAF-UG:-L5lYBvzmiVb0CjCGYhh:blrqfev) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4c5a20a88adfae4b2c389ec27c6da8cb643396/pkg/sql/opt/constraint/constraint_set_test.go#L204)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

What about also testing, for example, that `spGt20.Union(&evalCtx, spLe40)` is unconstrained?
</blockquote></details>

I've got a test above that does that: `// @1 > 20 OR @1 = 10 OR @1 <= 40`

---

*[pkg/sql/opt/constraint/constraint_set_test.go, line 222 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5jC-87ltLryVPJU5g5:-L5lYmcq0XUjszQt_cEx:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4c5a20a88adfae4b2c389ec27c6da8cb643396/pkg/sql/opt/constraint/constraint_set_test.go#L222)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

should this be [/80 - /80]?
</blockquote></details>

Done.

---

*[pkg/sql/opt/constraint/constraint_set_test.go, line 247 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5jBokRqXv81ttGeQOm:-L5lYnCp2T5nuMxYqXH-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4c5a20a88adfae4b2c389ec27c6da8cb643396/pkg/sql/opt/constraint/constraint_set_test.go#L247)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

should this be [/80 - /80]?
</blockquote></details>

Done.

---

*[pkg/sql/opt/constraint/constraint_test.go, line 38 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5jKeY3f-jEOFz4s-TM:-L5lYuKtzrkl-5UuEHfh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4c5a20a88adfae4b2c389ec27c6da8cb643396/pkg/sql/opt/constraint/constraint_test.go#L38)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

I think you should change `clone.String()` to `actual` 
</blockquote></details>

Done.

---

*[pkg/sql/opt/constraint/constraint_test.go, line 57 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5jNFOw0UKovQcI4bdY:-L5lZwqIZlOQSMRSimfB:bwukskr) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4c5a20a88adfae4b2c389ec27c6da8cb643396/pkg/sql/opt/constraint/constraint_test.go#L57)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

How is this "tight"?
</blockquote></details>

Not sure, I think I meant to create an "adjacent span" test here. Fixed to do that.

---

*[pkg/sql/opt/constraint/constraint_test.go, line 64 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5jQApN2fNmX2En7ATo:-L5laUYvHbRZnT7PaPF4:b2xn5k5) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4c5a20a88adfae4b2c389ec27c6da8cb643396/pkg/sql/opt/constraint/constraint_test.go#L64)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Why not have another helper function to call `tryUnionWith` and also check that the boolean output is correct?
</blockquote></details>

Well, I tried that, but I didn't like the result, because it made the code look as if I'm trying to test `tryUnionWith`, when in fact I'm only using it to set up the actual test. That made the entire test harder to read. In addition, it really added no additional testing value, since if the boolean output was incorrect, the final output would change anyway, and I'd catch the problem regardless.

---

*[pkg/sql/opt/constraint/constraint_test.go, line 118 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5jP2DsJix-tU9h2zZQ:-L5lbKgPf8tglHLzkCsR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4c5a20a88adfae4b2c389ec27c6da8cb643396/pkg/sql/opt/constraint/constraint_test.go#L118)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

I think you should change `clone.String()` to `actual` 
</blockquote></details>

Done.

---

*[pkg/sql/opt/constraint/constraint_test.go, line 175 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5jLA53DlOCTBWunAdX:-L5lbQK3YC0sy15yjn8p:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4c5a20a88adfae4b2c389ec27c6da8cb643396/pkg/sql/opt/constraint/constraint_test.go#L175)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

change this to [ - /10) and (/20 - ]
</blockquote></details>

Done.

---

*[pkg/sql/opt/constraint/constraint_test.go, line 183 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5jMj0QPNtwnORtqHEa:-L5lba9vGj0AYFSkSRL_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4c5a20a88adfae4b2c389ec27c6da8cb643396/pkg/sql/opt/constraint/constraint_test.go#L183)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

should be `/'mango'/false`
</blockquote></details>

Done.

---

*[pkg/sql/opt/constraint/constraint_test.go, line 207 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5jLGwu4VA9XVZTvJcg:-L5lbcw23qwIDqcaZ1Mx:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4c5a20a88adfae4b2c389ec27c6da8cb643396/pkg/sql/opt/constraint/constraint_test.go#L207)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

change this to [ - /10)
</blockquote></details>

Done.

---

*[pkg/sql/opt/constraint/constraint_test.go, line 211 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5jLHkdEukHeQjGXI3P:-L5lbeGJZGsQOV6Grs9M:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4c5a20a88adfae4b2c389ec27c6da8cb643396/pkg/sql/opt/constraint/constraint_test.go#L211)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

change this to (/20 - ]
</blockquote></details>

Done.

---

*[pkg/sql/opt/constraint/constraint_test.go, line 243 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5jMH7Kd9qcoMT-s9_A:-L5lbgk9Y1zeHr6uROgT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4c5a20a88adfae4b2c389ec27c6da8cb643396/pkg/sql/opt/constraint/constraint_test.go#L243)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

should be `/'mango'/false`
</blockquote></details>

Done.

---

*[pkg/sql/opt/constraint/key.go, line 193 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5jTR-f9_cl6Pz0P9Rz:-L5lbzLt29_UIO9y7OUF:by5blkf) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4c5a20a88adfae4b2c389ec27c6da8cb643396/pkg/sql/opt/constraint/key.go#L193)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Would be nice to move this up to the `KeyExtension` type definition so all the methods for the same type are together
</blockquote></details>

I decided to put `KeyExtension` in its own file.

---

*[pkg/sql/opt/constraint/span.go, line 45 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5jWpuM7Rf9IrowISDa:-L5lcDvNcIWrXAwI8Kb9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4c5a20a88adfae4b2c389ec27c6da8cb643396/pkg/sql/opt/constraint/span.go#L45)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Make the +inf and -inf square brackets
</blockquote></details>

Done.

---

*[pkg/sql/opt/constraint/span.go, line 224 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5jbN2aUtppZ9i6aR66:-L5ldOZo-nZe4ADzw42Y:b-rrjun1) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4c5a20a88adfae4b2c389ec27c6da8cb643396/pkg/sql/opt/constraint/span.go#L224)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Shouldn't this just be `<`, not `<=`?
</blockquote></details>

Well, it's fine if the equal case is handled there, but for clarity I changed this to `<` and changed the greater case to explicitly check for `>`.

---

*[pkg/sql/opt/constraint/span.go, line 229 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5jbItk6AfGS91WFQ3J:-L5ldJXJssS9B7ZKSpmH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4c5a20a88adfae4b2c389ec27c6da8cb643396/pkg/sql/opt/constraint/span.go#L229)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Should say this span is greater
</blockquote></details>

Done.

---

*[pkg/sql/opt/constraint/span.go, line 230 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5jbYf9excwAj9jSpjr:-L5ldFrR3bXnq-pHGn6t:b-zggshm) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4c5a20a88adfae4b2c389ec27c6da8cb643396/pkg/sql/opt/constraint/span.go#L230)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

...and before the start of...
</blockquote></details>

Actually, I think it's correct as-is, because it's "after the end of" the *other* span.

---

*[pkg/sql/opt/constraint/span_test.go, line 331 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5jgIaOQAEaWyhHTB3W:-L5lf_tTWR1jNduOmNlv:b-q82zuu) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4c5a20a88adfae4b2c389ec27c6da8cb643396/pkg/sql/opt/constraint/span_test.go#L331)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

What about testing intersection with `Span{}`?
</blockquote></details>

Good idea. I added a test here and in Union case.

---

*[pkg/sql/opt/constraint/span_test.go, line 421 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5jfYGC5VXli5uyN9Ks:-L5lfbi_BiIIP1t2kctg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4c5a20a88adfae4b2c389ec27c6da8cb643396/pkg/sql/opt/constraint/span_test.go#L421)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Did you mean to swap `pear` and `pear2` here?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
Thanks for the thorough reviews. I'm finding it can be harder to explain how the code works than to write the code itself!

---

Review status: 3 of 11 files reviewed at latest revision, 34 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-:-L5lg0k8A9GZ4u7fmJL5:b-jqv5gq)*
<!-- Sent from Reviewable.io -->

rytaft commented:
Haha yea - and there are no tests for comments so it's harder to catch the typos :) The new changes look great!

---

Reviewed 8 of 8 files at r4.
Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt/constraint/constraint.go, line 168 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5irFUW5u879_JQ2V6a:-L5nm-4ltCKqM7dEfibk:b-kczyc2) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4c5a20a88adfae4b2c389ec27c6da8cb643396/pkg/sql/opt/constraint/constraint.go#L168)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

You mean that the boolean return value indicates something very different? Both mean "the result of the union cannot be represented as a valid Span/Constraint". I added some commentary to that effect, but if you want to suggest something that further reduces the confusion, please do and I can add. I also changed the logic a bit to be more explicit about Unconstrained Span handling. Not allowing Unconstrained Constraints was something Radu felt was important, and that drives some of this logic.
</blockquote></details>

New comments are helpful - thanks!

---

*[pkg/sql/opt/constraint/constraint.go, line 210 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5nlUhxvNMHYPOVxvNd:-L5nlUhy4K-ECu80bUx8:bal2b21) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e5018a1f773ff4025bf66f9fc41ab0d0b0cb3a/pkg/sql/opt/constraint/constraint.go#L210)):*
> ```Go
> 	}
> 
> 	// We've got a valid constraint,
> ```

[nit] , -> .

---

*[pkg/sql/opt/constraint/constraint_test.go, line 64 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5jQApN2fNmX2En7ATo:-L5nnk6m0aulx4Lt-_FE:b2nhlkw) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4c5a20a88adfae4b2c389ec27c6da8cb643396/pkg/sql/opt/constraint/constraint_test.go#L64)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Well, I tried that, but I didn't like the result, because it made the code look as if I'm trying to test `tryUnionWith`, when in fact I'm only using it to set up the actual test. That made the entire test harder to read. In addition, it really added no additional testing value, since if the boolean output was incorrect, the final output would change anyway, and I'd catch the problem regardless.
</blockquote></details>

Ok!

---

*[pkg/sql/opt/constraint/span.go, line 230 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5jbYf9excwAj9jSpjr:-L5noOsmi-7Lw2s98sri:bu60mwt) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4c5a20a88adfae4b2c389ec27c6da8cb643396/pkg/sql/opt/constraint/span.go#L230)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Actually, I think it's correct as-is, because it's "after the end of" the *other* span.
</blockquote></details>

Right, but I was thinking the sentence would be clearer if it followed the structure of the sentence in the previous block:

```
// This span is greater, so see if there's any "space" after the end of 
// the other span and before the start of this span.
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-:-L5np932DZLLZzUvPUHu:bqrahn2)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt/constraint/constraint.go, line 210 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5nlUhxvNMHYPOVxvNd:-L5nuPuvs1mxs0oLT6I6:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5e5018a1f773ff4025bf66f9fc41ab0d0b0cb3a/pkg/sql/opt/constraint/constraint.go#L210)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

[nit] , -> .
</blockquote></details>

Done.

---

*[pkg/sql/opt/constraint/span.go, line 230 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768#-L5jbYf9excwAj9jSpjr:-L5nu2NFAIOg2PzosMEQ:bwfpptp) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa4c5a20a88adfae4b2c389ec27c6da8cb643396/pkg/sql/opt/constraint/span.go#L230)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Right, but I was thinking the sentence would be clearer if it followed the structure of the sentence in the previous block:

```
// This span is greater, so see if there's any "space" after the end of 
// the other span and before the start of this span.
```
</blockquote></details>

Ah I see. Changed it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22768)*
<!-- Sent from Reviewable.io -->
